trlx/orchestrator/offline_orchestrator.py

```diff
@@ -10,67 +53,84 @@ class OfflineOrchestrator(Orchestrator):
     Orchestrator that creates a static dataset for offline training
     """
 
-    def __init__(self, trainer, split_token=None):
+    def __init__(self, trainer):
         self.trainer = trainer
-        self.split_token = split_token
 
-    def make_experience(self, samples, rewards):
+    def make_experience(self, samples, rewards, max_length=2048):
         """
-        Tokenizes samples and shapes rewards into proper tensors and then inserts the resulting dataset into the model
+        Tokenizes samples and shapes rewards into proper tensors and then inserts the resulting dataset into the trainer
         """
         if self.trainer.tokenizer:
-            input_ids = self.trainer.tokenize(samples)
-        else:
-            input_ids = samples
-
-        input_ids = list(map(torch.as_tensor, input_ids))
-
-        states_ixs, actions_ixs = [], []
-        dones = []
-        for s, s_tok in zip(samples, input_ids):
-            # split samples on (prompts, continuations) on a given substring `split_token`
-            if self.split_token:
-                prompt_str_len = s.index(self.split_token) + len(self.split_token)
-                prompt_tok_len = len(
-                    self.trainer.tokenizer(s[:prompt_str_len]).input_ids
+            samples = [
+                tokenize_dialogue(
+                    s, self.trainer.tokenizer, max_length, truncation_side="right"
                 )
-            # else assume that the prompt is a bos token
-            else:
-                prompt_tok_len = 1
-
-            # indices of continuations, to mask prompts in loss computation
-            a_ixs = torch.arange(prompt_tok_len - 1, len(s_tok) - 1)
-            # same continuations but for value computation, with the premise to eventually support interleaved dialog
-            s_ixs = torch.arange(prompt_tok_len - 1, len(s_tok))
-            # mask continuation's ending
-            terminals = torch.ones_like(s_ixs)
-            terminals[-1] = 0
-
-            actions_ixs.append(a_ixs)
-            states_ixs.append(s_ixs)
-            dones.append(terminals)
+                for s in samples
+            ]
+
+        all_input_ids = []
+        all_actions_ixs = []
+        all_states_ixs = []
+        all_dones = []
+        for sample in samples:
+            length = 0
+            all_input_ids.append(torch.tensor(sum(sample, [])))
+            isoutput = False
+            actions_ixs, states_ixs = [], []
+            for phrase in sample:
```

Comment:
So a sample can contain multiple "phrases" which we can reward?

Is there a reason we can't just separate each phrase into it's own datapoint with the "output" at the end(with all the prior dialog as context)? Do you see some advantage to treating things this way?

Reply:
Yes, both are supported and opt-in